cli/src/diff_util.rs

```diff
@@ -346,19 +347,66 @@ fn show_color_words_diff_line(
     Ok(())
 }
 
-fn diff_content(path: &RepoPath, value: MaterializedTreeValue) -> Result<Vec<u8>, CommandError> {
+struct FileContent {
+    /// false if this file is likely text; true if it is definitely binary.
+    is_binary: bool,
+    // TODO: maybe this could store an `enum { String, Vec<u8> }` instead? but then we have to
+    // UTF-8 validate the whole file.
+    contents: Vec<u8>,
+}
+
+impl FileContent {
+    pub(crate) fn is_empty(&self) -> bool {
+        self.contents.is_empty()
+    }
+}
+
+impl From<String> for FileContent {
+    fn from(contents: String) -> Self {
+        FileContent {
+            is_binary: false,
+            contents: contents.into_bytes(),
+        }
+    }
+}
+
+fn file_content_for_diff(reader: &mut dyn io::Read) -> io::Result<FileContent> {
+    // If this is a binary file, don't show the full contents.
+    // Determine whether it's binary by whether the first 8k bytes contain a null
+    // character; this is the same heuristic used by git as of writing: https://github.com/git/git/blob/eea0e59ffbed6e33d171ace5be13cde9faa41639/xdiff-interface.c#L192-L198
+    const PEEK_SIZE: usize = 8000;
+    // TODO: currently we look at the whole file, even though for binary files we
+    // only need to know the file size. To change that we'd have to extend all
+    // the data backends to support getting the length.
+    let mut contents = vec![];
+    reader.read_to_end(&mut contents)?;
+
+    let start = &contents[..PEEK_SIZE.min(contents.len())];
+    let is_binary = !start.is_utf8() || start.contains(&b'\0');
```

Comment:
Better to remove `.is_utf8()`. There are various encodings that aren't binary nor UTF-8. For example, log files sometimes contain arbitrary bytes, but diffing them is useful.

Reply:
I'll defer to Yuya on this. I don't have a strong opinion. I never needed to diff non-UTF8, but that doesn't necessarily mean there's never a good reason for it.

Reply:
hmm, i have mixed feelings about this; i think in practice almost all diff tools will handle non-utf8 poorly.

that said i don't have strong opinions, and it does seem good to be consistent with git.

Reply:
If you're just looking for a tie-breaker: I vote for doing what git does (i.e. just looking for null bytes, IIUC) :)

Reply:
Traditional tools (`diff`, `hg`, `git`, etc.) have no problem with non-utf8 data so long as the encoding is superset of ascii.

It would be annoying if `jj diff` failed just because the file contains copyright comment in latin1 encoding.